src/lib/merkle_mask/masking_merkle_tree.ml

```diff
@@ -181,7 +181,16 @@ module Make (Inputs : Inputs_intf.S) = struct
       | Some account ->
           Some account
       | None ->
-          Base.get (get_parent t) location
+          let is_empty =
+            match t.current_location with
```

Comment:
```suggestion
            match (last_filled t) with
```

`(last_filled t)` should probably be used here ? to get updated location from all the parents.
This would allow to keep the test above to pass.

Reply:
I'll defer to @nholland94, but IMO we break a bunch of invariants (e.g. merkle paths) if we allow incompatible writes to the parent ledger, and the test is misguided.

Reply:
This affects masks behavior: that would break this property of masks, which is quite important IMO:
> Whenever a read occurs on a mask, the mask will first attempt to retrieve the node from its internal sparse collection, and if it does not exist there, will forward the read request to its parent.

(text from [specs/merkle_tree.md](https://github.com/MinaProtocol/mina/blob/develop/docs/specs/merkle_tree.md))

If the performance of `(last_filled t)`  is the concern, it could be called only once in `get_batch` and `self_merkle_path`

Reply:
Hmm, this is confusing. We should probably update the mask interface to not allow updating parent masks if it breaks the invariants. I'm seeing calls to `parent_set_notify` which is to notify child masks about updates in the parent mask?
Re the test, if the dummy account was added to the parent mask before creating the child mask, then it should pass?

Reply:
> We should probably update the mask interface to not allow updating parent masks if it breaks the invariants.

This was my suggestion last week, but I believe consensus was that this will be too big of a change to implement right now.

But in general I agree with this, though supposedly related functionalities are just dead code (the riskiness of a change lies in the fact that it'd require a complicated analysis of the whole codebase to prove it's actually dead, though I'd argue seeing it alive is worse and we'd better throw a runtime error when such things occur).

Reply:
Ok. created #14631 to track this. We should address this before making any planned changes to this library